### Heart Attack Prediction - Coursework 1

#### Problem Definition

Heart attacks are a serious condition and a life-threatening emergency. Artificial Intelligence can help predict heart attacks and save lives. 
We are looking for an AI-based model to predict the heart attack given some health indicators such as BMI, age, being a smoker, with high accuracy. <br>
Please write a report in 1500 words explaining in detail different stages of achieving the goal of this project. The report should include the following steps in detail. The dataset has been provided for you on Moodle. <br>
1.1 Prepare the dataset for training (Check for missing values, splitting the dataset in training and testing datasets…). <br>
1.2 Choose and tailor 4 Models you have learnt during lectures including an ANN-based model and train and build the models. <br>
1.3 Evaluate your model using metrics you have learnt during the lectures. <br>
1.4 Compare the models and report the most accurate model for this purpose. <br>
1.5 Please provide 2 ways (with evidence) to further improve the accuracy of your best model.

#### Project Goal

- To create a model that can predict whether someone will get a heart attack or not

Project sections:
1. Data ingestion, Data preprocessing, Data segragation
2. Model selection, Model training
3. Model evaluation (use all metrics) - recall is the most important
4. Compare selected models and report the most accurate model
5. Provide 2 ways (with evidence) to further improve the accuracy of the model selected in (4)

#### Importing Required Python Libraries

In [2]:
import pandas as pd

#### Section 1

##### 1.1 Data Ingestion

In [34]:
heart_data = pd.read_csv("heart_attack.csv")
heart_data.head()

,HeartDiseaseorAttack,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,Diabetes,PhysActivity,HvyAlcoholConsump,MentHlth,PhysHlth,Sex,Age,Education,Income
0,0,1,1,1,40.0,1,0,0,0,0,18,15,0,9,4,3
1,0,0,0,0,25.0,1,0,0,1,0,0,0,0,7,6,1
2,0,1,1,1,28.0,0,0,0,0,0,30,30,0,9,4,8
3,0,1,0,1,27.0,0,0,0,1,0,0,0,0,11,3,6
4,0,1,1,1,24.0,0,0,0,1,0,3,0,0,11,5,4


##### 1.2 Data Insights/Analysis

In [35]:
# Every feature is of integer type, so encoding is not necessary for this dataset
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129998 entries, 0 to 129997
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   HeartDiseaseorAttack  129998 non-null  int64  
 1   HighBP                129998 non-null  int64  
 2   HighChol              129998 non-null  int64  
 3   CholCheck             129998 non-null  int64  
 4   BMI                   129988 non-null  float64
 5   Smoker                129998 non-null  int64  
 6   Stroke                129998 non-null  int64  
 7   Diabetes              129998 non-null  int64  
 8   PhysActivity          129998 non-null  int64  
 9   HvyAlcoholConsump     129998 non-null  int64  
 10  MentHlth              129998 non-null  int64  
 11  PhysHlth              129998 non-null  int64  
 12  Sex                   129998 non-null  int64  
 13  Age                   129998 non-null  int64  
 14  Education             129998 non-null  int64  
 15  

In [8]:
# statistical analysis of the data
heart_data.describe().T

,count,mean,std,min,25%,50%,75%,max
HeartDiseaseorAttack,129998.0,0.093509,0.291146,0.0,0.0,0.0,0.0,1.0
HighBP,129998.0,0.429614,0.495023,0.0,0.0,0.0,1.0,1.0
HighChol,129998.0,0.426322,0.494544,0.0,0.0,0.0,1.0,1.0
CholCheck,129998.0,0.963315,0.187988,0.0,1.0,1.0,1.0,1.0
BMI,129988.0,28.435586,6.999582,12.0,24.0,27.0,31.0,98.0
Smoker,129998.0,0.448030,0.497294,0.0,0.0,0.0,1.0,1.0
Stroke,129998.0,0.041354,0.199110,0.0,0.0,0.0,0.0,1.0
Diabetes,129998.0,0.296689,0.697608,0.0,0.0,0.0,0.0,2.0
PhysActivity,129998.0,0.760758,0.426622,0.0,1.0,1.0,1.0,1.0
HvyAlcoholConsump,129998.0,0.057562,0.232915,0.0,0.0,0.0,0.0,1.0


Points I noted from the stats above;
1. Only 9.35% of people had a heart attack according to the dataset, hence there's a bias towards people not getting heart attacks
2. 96% of people went for cholesterol check
3. The highest BMI value is 98, these are likely outliers and might be dropped
4. 76% of people are physically active
5. Only 5.7% of people are heavy alcohol consumers
6. The average age group of the dataset is between 60-64 (age group 8)
7. BMI, Diabetes, MentHlth, PhysHlth, Age, Education, and Income are not binary values

In [36]:
# checking for null values
heart_data.isnull().sum()

HeartDiseaseorAttack     0
HighBP                   0
HighChol                 0
CholCheck                0
BMI                     10
Smoker                   0
Stroke                   0
Diabetes                 0
PhysActivity             0
HvyAlcoholConsump        0
MentHlth                 0
PhysHlth                 0
Sex                      0
Age                      0
Education                0
Income                   0
dtype: int64

Points I noted from the null values output;
1. BMI has 10 null values